x-pack/metricbeat/module/aws/awshealth/awshealth.go

```diff
@@ -0,0 +1,448 @@
+func createEvents(hd HealthDetails) mb.Event {
+	currentDate := getCurrentDateTime()
+	//eventID := currentDate + getStringValueOrDefault(hd.event.Arn) + getStringValueOrDefault((*string)(&hd.event.StatusCode))
+	eventID := currentDate + getValueOrDefault(hd.event.Arn, "") + getValueOrDefault((*string)(&hd.event.StatusCode), "")
+	event := mb.Event{
+		MetricSetFields: mapstr.M{
+			"event_arn":                  getValueOrDefault(hd.event.Arn, ""),
+			"end_time":                   getValueOrDefault(hd.event.EndTime, time.Time{}),
+			"event_scope_code":           getValueOrDefault((*string)(&hd.event.EventScopeCode), ""),
+			"event_type_category":        getValueOrDefault((*string)(&hd.event.EventTypeCategory), ""),
+			"event_type_code":            getValueOrDefault(hd.event.EventTypeCode, ""),
+			"last_updated_time":          getValueOrDefault(hd.event.LastUpdatedTime, time.Time{}),
+			"region":                     getValueOrDefault(hd.event.Region, ""),
+			"service":                    getValueOrDefault(hd.event.Service, ""),
+			"start_time":                 getValueOrDefault(hd.event.StartTime, time.Time{}),
+			"status_code":                getValueOrDefault((*string)(&hd.event.StatusCode), ""),
+			"affected_entities_pending":  hd.affectedEntityPending,
+			"affected_entities_resolved": hd.affectedEntityResolved,
+			"affected_entities_others":   hd.affectedEntityOthers,
+			"affected_entities":          createAffectedEntityDetails(hd.affectedEntities),
+		},
+		RootFields: mapstr.M{
+			"cloud.provider": "aws",
+		},
+		ID: generateEventID(eventID),
+	}
+	return event
+}
+
+type HealthDetails struct {
+	event                  types.Event
+	eventDescription       string
+	affectedEntities       []types.AffectedEntity
+	affectedEntityPending  int32
+	affectedEntityResolved int32
+	affectedEntityOthers   int32
+}
+
+type AffectedEntityDetails struct {
+	AwsAccountId    string    `json:"aws_account_id"`
+	EntityUrl       string    `json:"entity_url"`
+	EntityValue     string    `json:"entity_value"`
+	LastUpdatedTime time.Time `json:"last_updated_time"`
+	StatusCode      string    `json:"status_code"`
+	EntityArn       string    `json:"entity_arn"`
+}
+
+// getValueOrDefault returns the dereferenced value of a pointer v of any type T.
+// If the pointer is nil, it returns the specified defaultValue of type T.
+func getValueOrDefault[T any](v *T, defaultValue T) T {
+	if v != nil {
+		return *v
+	}
+	return defaultValue
+}
+
+// createAffectedEntityDetails populates and returns a slice of AffectedEntityDetails
+// based on the given list of AffectedEntity instances.
+// Each AffectedEntity is converted into an AffectedEntityDetails struct.
+func createAffectedEntityDetails(affectedEntities []types.AffectedEntity) []AffectedEntityDetails {
+	aed := []AffectedEntityDetails{}
+	// Populate a slice of AffectedEntityDetails
+	for _, entity := range affectedEntities {
```

Comment:
It is safer to use indexing or copy of entity:

> entity := entity // or affectedEntities[i]

Issue: Implicit memory aliasing because of usage of append and as I told you earlier for range creates a copy to a fixed memory location that is updated every iteration.

Although, in Go 1.22, no problem will be there. But ideally, we should fix this.

https://tip.golang.org/doc/go1.22#:~:text=Go%201.22%20makes%20two%20changes%20to%20%22for%22%20loops.